src/parser/core/modules/CooldownDowntime.js

```diff
@@ -0,0 +1,183 @@
+	_downtimeOkTime = 0
+	//Determines if we're being overriden by an extension class
+	_extensionOverride = false
+
+	constructor(...args) {
+		super(...args)
+
+		this.addHook('begincast', {by: 'player'}, this._onBegin)
+		this.addHook('cast', {by: 'player'}, this._onCast)
+		this.addHook('complete', this._onComplete)
+	}
+
+	/**
+	 *This just sets the internal ability list if you aren't going to use override functions and
+	 perform calculated data holding in your extended class
+	 * @param {List of Abilities to track} abilities
+	 */
+	setAbilityList(abilities) {
+		this._trackedcds = abilities
+	}
+
+	/**
+	 * This instructs the module how long a CD can be held for before it is
+	 * problematic to hold.
+	 * @param {Time in MS allowed to hold a CD for} duration
+	 */
+	setDownTimeDuration(duration) {
+		this._downtimeOkTime = duration
+	}
+
+	/**
+	 * Denotes if we're overriding this super from an extension
+	 * @param {boolean, whether or not we override this super from an extension} override
+	 */
+	setExtensionOverride(override) {
+		this._extensionOverride = override
+	}
+
+	_onBegin(event) {
+
+		//do the check on begin (even tho they don't have a cast time)
+		this._cooldownCheck(event)
+	}
+
+	_onCast(event) {
+		//do the check at every cast
+		this._cooldownCheck(event)
+	}
+
+	_onComplete(event) {
+		const endTime = event.timestamp
+		const startTime = this.parser.fight.start_time
+		const encounterLength = endTime - startTime
+		const OGCDRequirements = []
+		Object.keys(this._downTime).forEach(id => {
+			const dt = this._downTime[id]
+
+			//wrap up all the open ones and save them in history
+			if (dt.current) {
+				dt.current.stoptime = endTime
+				dt.history.push(dt.current)
+				dt.current = null
+			}
+
+			//calculate the downtime based on the start and stop values and sum the array
+			//Adjust for the classes defined alloted time to allow a CD to be held
+			//this supports classes like RDMs who routinely hold CDs due to procs
+			const totalSumOfDownTime = dt.history.map(downTime => {
+				return this._getDownTimeAdjustedForInvuln(downTime.starttime, downTime.stoptime)
+				//return Math.max(downTime.stoptime - downTime.starttime - this._downtimeOkTime, 0)
+			}).reduce(
+				(accumulator, currentValue) => accumulator + currentValue
+			)
+			//write the results as a new Requirement to show up later
+			OGCDRequirements.push(
+				new Requirement({
+					name: <ActionLink {...getAction(id)} />,
+					percent: this._percentFunction(id, totalSumOfDownTime, encounterLength),
+				})
+			)
+
+		})
+
+		//new Rule and adds the array of Requirements that just got generated
+		this.checklist.add(new Rule({
+			name: 'Use your OGCDs',
+			description: `Always make sure to use your OGCDs when they are up but don't clip them.  ${this._downtimeOkTime === 0 ? '' : `To account for random factors you are given a buffer of ${this._downtimeOkTime/1000} seconds per instance to use your CD`}`,
```

Comment:
I think this wording could do with some work but I'm by no means a master of wording :blobsweat:.

Reply:
Heh this is iteration of wording like 20; I'm not a master of wording either